### PR TITLE
Rst well shut2

### PR DESCRIPTION
**Pull request by joakim-hove**
Have just extracted the important commits from #1656

This quite small PR will require much data updates .... good if someone takes a hard look.

### CODE REVIEW

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

File: src/opm/parser/eclipse/EclipseState/Schedule/Well/Well.cpp [hunk 1]
```diff
@@ -179,36 +179,39 @@ Well::Well(const RestartIO::RstWell& rst_well,
...
+            case 6:
+                p->controlMode = Well::ProducerCMode::THP;
+                p->addProductionControl(Well::ProducerCMode::THP);
+                break;
+            case 7:
+                p->controlMode = Well::ProducerCMode::BHP;
+                p->addProductionControl(Well::ProducerCMode::BHP);
+                break;
+            default:
+                throw std::invalid_argument("Can convert integer value: " + std::to_string(rst_well.active_control)
```
Comment (bska) [inline src/opm/parser/eclipse/EclipseState/Schedule/Well/Well.cpp]: Is this supposed to say
```
can NOT convert …
```
?

File: src/opm/parser/eclipse/EclipseState/Schedule/Well/Well.cpp [hunk 2]
```diff
@@ -1376,6 +1382,9 @@ int Opm::eclipseControlMode(const Opm::Well::InjectorCMode imode,
 
     using IType = ::Opm::InjectorType;
 
+    if (wellStatus == ::Opm::Well::Status::SHUT) {
+        return Val::Shut;
+    }
```
Comment (bska) [inline src/opm/parser/eclipse/EclipseState/Schedule/Well/Well.cpp]: This is going to require updating [`WellTests`](https://github.com/OPM/opm-common/blob/269537b1015aedd904e55dc26ba18091bf25bf1d/tests/parser/WellTests.cpp#L927) too.

File: src/opm/parser/eclipse/EclipseState/Schedule/Well/Well.cpp [hunk 3]
```diff
@@ -1406,6 +1415,9 @@ int Opm::eclipseControlMode(const Opm::Well::ProducerCMode pmode,
     using PMode = ::Opm::Well::ProducerCMode;
     using Val   = ::Opm::RestartIO::Helpers::VectorItems::IWell::Value::WellCtrlMode;
 
+    if (wellStatus == ::Opm::Well::Status::SHUT) {
+        return Val::Shut;
+    }
```
Comment (bska) [inline src/opm/parser/eclipse/EclipseState/Schedule/Well/Well.cpp]: This too requires updating [`WellTests`](https://github.com/OPM/opm-common/blob/269537b1015aedd904e55dc26ba18091bf25bf1d/tests/parser/WellTests.cpp#L1038).

**Comment by joakim-hove** [comment]
> As an example, this PR means that a producing well won't support THP controls in a restarted run unless the well was open and constrained by the THP limit at the time of the restart.

Yes - that is a consequence of this PR; and that is how I understand the information actually encoded in the restart file; I am more than open to corrections if you think I have got it wrong.

File: src/opm/parser/eclipse/EclipseState/Schedule/Well/Well.cpp (hunk 1 above)
Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/Schedule/Well/Well.cpp]: Ehhh - yes :-)

**Comment by bska** [comment]
>> As an example, this PR means that a producing well won't support THP controls in a restarted run unless the well was open and constrained by the THP limit at the time of the restart.
> 
> Yes - that is a consequence of this PR; and that is how I understand the information actually encoded in the restart file; I am more than open to corrections if you think I have got it wrong.

I don't see how that's appropriate.  It would be akin to discarding the THP limit when processing one of the `WCON*` keywords in an input file.  The well would never be able to switch to THP control.  Essentially, `SWEL` is a simplified/linearised copy of the input data&mdash;especially the targets and limits&mdash;whereas `XWEL` is the simulator's solution state for each well.

If there truly is no THP limit then the corresponding SWEL item will have the sentinel value `1.0e+20`.

**Comment by joakim-hove** [comment]
> I don't see how that's appropriate. It would be akin to discarding the THP limit when processing one of the WCON* keywords in an input file.

OK; here at least I misunderstand things - it was never my intention to remove functionality. I will look over it again and get back to you tomorrow if I still can't get my head around it.

**Comment by joakim-hove** [comment]
> this PR means that a producing well won't support THP controls in a restarted run unless the well was open and constrained by the THP limit at the time of the restart.

Assume the following:

1. The well is producing under THP constraint.
2. The well is closed.
3. We write a restart file while the well is closed.

When we load this well from the restart file we will have `IWEL[Ix::ActWCtrl] == 6` - corresponding to an active THP control. That is to some extent correct, but even more so the well is actually closed. As I have understood it[1] the way to indicate that a well is closed is by writing `IWEL[Ix::ActWCtrl] == 0` to the restart file - and that is the main change in this PR; so in that sense this PR will remove "active" THP control, but it will not remove the ability to again go under THP control if the well reopens.

I'll contact you on &phone; later today

[1]: I would have thought that the OPEN/SHUT status had a flag of it's own; that might be the fundamental misunderstanding on my part.

**Comment by joakim-hove** [comment]
OK - the `THP` case has been fixed; tests updated. Now this should be ready!

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by bska** [comment]
> OK - the THP case has been fixed; tests updated. Now this should be ready!

Cool.  I'll review the code and the test failures and if everything is as expected I'll update the reference data.

**Comment by bska** [comment]
jenkins build this update_data please

**Comment by joakim-hove** [comment]
> I've verified that the test failures are also all of the expected kind

Thanks - that is a quite laborious manual task.

**Comment by bska** [comment]
jenkins build this opm-tests=276 please

**Comment by bska** [comment]
Goes through without issue with updated reference data from OPM/opm-tests#276.  I'm going to merge this into master now.  Thanks a lot for the updates.

**Pull request merged**